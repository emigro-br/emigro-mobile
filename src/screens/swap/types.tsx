
export enum SwapType {
   SELL,
   BUY
}
