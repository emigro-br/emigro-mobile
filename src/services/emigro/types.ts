//-- Auth
export const enum Role {
  VENDOR = 'VENDOR',
  CUSTOMER = 'CUSTOMER',
}

export type UserCredential = {
  session: AuthSession;
  user: User;
};

export type AuthSession = {
  accessToken: string;
  refreshToken: string;
  idToken: string;
  tokenExpirationDate: Date;
};

export type ConfirmUserRequest = {
  externalId: string;
  email: string;
  code: string;
};

export type RegisterUserRequest = {
  email: string;
  password: string;
  firstName: string;
  lastName: string;
  role: Role;
  [key: string]: string;
};

//-- Users

export type User = {
  id: string;
  externalId: string;
  publicKey: string;
  secretKey: string;
  role: Role;
  status: string;
  createdAt: string;
  updatedAt: string;
  preferences: Record<string, any>;
};

export type UserProfile = {
  sub: string;
  address: string;
  email_verified: string;
  given_name: string;
  family_name: string;
  custom_role: string;
  email: string;
};

export type StellarAccount = {
  publicKey: string;
  secretKey?: string;
  balances: Balance[];
};

export type Balance = {
  balance: string; // TODO: convert to number
  assetType: string;
  assetCode: string;
};

//-- Anchors

export type InteractiveUrlRequest = {
  asset_code: string;
  // keep this type, we can add more fields in the future
};

export type InteractiveUrlResponse = {
  id: string; // anchor transactionId
  type: string;
  url: string;
};

//-- Transactions

export type TransactionType =
  | 'transfer'
  | 'payment'
  | 'withdrawal'
  | 'deposit'
  | 'swap'
  | 'exchange'
  | 'fee'
  | 'refund'
  | 'other';

export type CreateTransactionRequest = {
  type: TransactionType;
  maxAmountToSend: string | undefined;
  destinationAmount: string;
  destination: string;
  sourceAssetCode: string;
  destinationAssetCode: string;
  idempotencyKey?: string;
};

export type CreateSwapTransaction = {
  fromAsset: string;
  toAsset: string;
  amount: number;
  estimated: number;
  idempotencyKey?: string;
};

export type CreateTransferTransaction = {
  destinationAddress: string;
  assetCode: string;
  amount: number;
  idempotencyKey?: string;
};

export type CreatePaymentTransaction = {
  destinationAddress: string;
  sendAssetCode: string;
  destAssetCode: string;
  destAmount: number;
  sendMax: number;
  idempotencyKey?: string;
};

export type Transaction = {
  id: string;
  type: string;
  from: string;
  to: string;
  amount: number;
  status: string;
  createdAt: Date;
};

// DERECATED: old transaction response
export type PaymentResponse = {
  transactionHash: string;
  sourceAccount: string;
  type: string;
  assetType: string;
  assetCode: string;
  assetIssuer: string;
  from: string;
  to: string;
  amount: string;
  path: string[];
  sourceAmount: string;
  sourceAssetType: string;
  sourceAssetCode: string;
  sourceAssetIssuer: string;
};

// PIX

export type BrcodePaymentRequest = {
  brcode: string;
  exchangeAsset: string;
  amount: number;
  taxId: string;
  description: string;
};

export type BrcodePaymentResponse = {
  id: string;
  brcode: string;
  taxId: string;
  description: string;
  amount: number;
  status: string;
};

export type DictKey = {
  pixKey: string;
  taxId: string;
  bankName: string;
};

export type PixPaymentPreview = {
  brcode: string;
  pixKey: string;
  amount: number;
  currency?: string;
  bankName: string;
  name: string;
  taxId: string;
  txId: string;
};

//-- Stellar Network
export enum Sep24TransactionStatus {
  COMPLETED = 'completed',
  ERROR = 'error',
  INCOMPLETE = 'incomplete',
  NON_INTERACTIVE_CUSTOMER_INFO_NEEDED = 'non_interactive_customer_info_needed',
  PENDING_ANCHOR = 'pending_anchor',
  PENDING_CUSTOMER_INFO_UPDATE = 'pending_customer_info_update',
  PENDING_EXTERNAL = 'pending_external',
  PENDING_RECEIVER = 'pending_receiver',
  PENDING_SENDER = 'pending_sender',
  PENDING_STELLAR = 'pending_stellar',
  PENDING_TRANSACTION_INFO_UPDATE = 'pending_transaction_info_update',
  PENDING_TRUST = 'pending_trust',
  PENDING_USER = 'pending_user',
  PENDING_USER_TRANSFER_START = 'pending_user_transfer_start',
}

export type Sep24Transaction = {
  amount_fee: string;
  amount_in: string;
  amount_out: string;
  completed_at: string | null;
  external_transaction_id: string;
  from: string;
  id: string;
  kind: string;
  message: string;
  more_info_url: string;
  refunded: boolean;
  started_at: string;
  status: Sep24TransactionStatus;
  status_eta: string | null;
  stellar_transaction_id: string | null;
  to: string | null;
  withdraw_anchor_account: string;
  withdraw_memo: string;
  withdraw_memo_type: string;
};
