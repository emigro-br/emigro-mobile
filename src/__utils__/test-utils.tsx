import { JSX, JSXElementConstructor, ReactElement } from 'react';

import { RenderOptions, render } from '@testing-library/react-native';

import { ThemeProvider } from './ThemeProvider';

type Props = {
  children: JSX.Element;
};

const AllTheProviders = ({ children }: Props) => {
  return <ThemeProvider theme="light">{children}</ThemeProvider>;
};

const customRender = (
  ui: ReactElement<unknown, string | JSXElementConstructor<any>>,
  options?: RenderOptions | undefined,
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react-native';

export { customRender as render };
