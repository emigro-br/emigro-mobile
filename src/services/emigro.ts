import { IBalance } from '@/types/IBalance';
import { IPaymentResponse } from '@/types/IPaymentResponse';
import { IQuote } from '@/types/IQuote';
import { IQuoteRequest } from '@/types/IQuoteRequest';
import { ITransaction } from '@/types/ITransaction';
import { ITransactionRequest } from '@/types/ITransactionRequest';
import { IUserProfile } from '@/types/IUserProfile';
import { refresh as refreshSession } from '@/services/auth';
import { getSession, saveSession } from '@/storage/helpers';

import { GET_USER_BALANCE_ERROR, QUOTE_NOT_AVAILABLE_ERROR, TRANSACTION_ERROR_MESSAGE } from '@constants/errorMessages';

const backendUrl = process.env.EXPO_PUBLIC_BACKEND_URL;


export class NotAuhtorized extends Error {
  constructor() {
    super('Not authorized');
  }
}

export const fetchWithTokenCheck = async (url: string, options: RequestInit): Promise<Response> => {

  let session = await getSession();
  if (!session) {
    throw new NotAuhtorized();
  }

  const { tokenExpirationDate } = session;
  const isTokenExpired = tokenExpirationDate < new Date();

  if (isTokenExpired) {
    console.debug('Token expired, refreshing...');
    const newSession = await refreshSession(session);
    if (newSession) {
      saveSession(newSession);
      session = newSession;
    } else {
      throw new Error('Could not refresh session');
    }
  }

  const { accessToken } = session;
  options.headers = {
    ...options.headers,
    Authorization: `Bearer ${accessToken}`,
  };

  return fetch(url, options);
}

export const getTransactions = async (): Promise<ITransaction[]> => {
  const transactionsUrl = `${backendUrl}/transaction/all`;

  try {
    const response = await fetchWithTokenCheck(transactionsUrl, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    });
    const { transactions } = await response.json();
    return transactions;
  } catch (error) {
    console.error(error);
    throw new Error();
  }
};

export const getUserBalance = async (): Promise<IBalance[]> => {
  const url = `${backendUrl}/user`;
  try {
    const response = await fetchWithTokenCheck(url, {
      method: 'GET',
    });
    const json = await response.json();

    if (json.statusCode === 401) {
      throw new NotAuhtorized();
    }

    const { balances } = json;
    return balances;
  } catch (error) {
    console.error(error, GET_USER_BALANCE_ERROR);
    throw new Error();
  }
};

export const handleQuote = async (body: IQuoteRequest): Promise<IQuote> => {
  const url = `${backendUrl}/quote`;

  try {
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
    const { quote } = await response.json();
    return quote;
  } catch (error) {
    console.error(error, QUOTE_NOT_AVAILABLE_ERROR);
    throw new Error();
  }
};

export const sendTransaction = async (transactionRequest: ITransactionRequest): Promise<IPaymentResponse> => {
  const url = `${backendUrl}/transaction`;
  try {
    const response = await fetchWithTokenCheck(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(transactionRequest),
    });
    return await response.json();
  } catch (error) {
    console.error(error, TRANSACTION_ERROR_MESSAGE);
    throw new Error();
  }
};

export const getUserPublicKey = async (): Promise<string> => {
  const url = `${backendUrl}/user`;

  try {
    const request = await fetchWithTokenCheck(url, {
      method: 'GET',
    });
    const { publicKey } = await request.json();
    return publicKey;
  } catch (error) {
    console.error(error, GET_USER_BALANCE_ERROR);
    throw error;
  }
};

export const getUserProfile = async (): Promise<IUserProfile> => {
  const url = `${backendUrl}/user/profile`;
  try {
    const session = await getSession();
    const response = await fetchWithTokenCheck(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(session),
    });
    const json = await response.json();
    if (json.statusCode === 401) {
      throw new NotAuhtorized();
    }
    return json
  } catch (error) {
    console.error(error);
    throw error;
  }
};
