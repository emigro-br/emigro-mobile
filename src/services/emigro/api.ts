import axios, { AxiosInstance, CreateAxiosDefaults } from 'axios';

import { CustomError } from '@/types/errors';

import { AuthSession } from './types';

export const backendUrl = process.env.EXPO_PUBLIC_BACKEND_URL;
const defaultTimeout = 10000;

export const api = (config: CreateAxiosDefaults = {}): AxiosInstance => {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const { sessionStore } = require('@/stores/SessionStore'); // workaround to avoid circular dependency

  const instance = axios.create({
    baseURL: backendUrl,
    timeout: defaultTimeout,
    ...config,
  });

  // Alter defaults after instance has been created
  if (sessionStore.session) {
    instance.defaults.headers.common['Authorization'] = `Bearer ${sessionStore.accessToken}`;
  }

  withRefreshTokenInterceptor(instance, sessionStore.refresh);

  if (process.env.DEBUG === 'axios') {
    withDebug(instance);
  }

  return instance;
};

const withDebug = (instance: AxiosInstance) => {
  instance.interceptors.request.use((request) => {
    console.debug('Starting Request', JSON.stringify(request, null, 2));
    return request;
  });

  instance.interceptors.response.use((response) => {
    console.debug('Response:', JSON.stringify(response, null, 2));
    return response;
  });
};

export function withRefreshTokenInterceptor(instance: AxiosInstance, refreshFn: () => Promise<AuthSession>) {
  let isRefreshing = false;

  instance.interceptors.response.use(
    (response) => response,
    async (error) => {
      const { config, response } = error;
      if (!response) {
        throw error;
      }

      const originalRequest = config;

      const Unauthorized = 401;
      if (response.status === Unauthorized && !originalRequest._retry) {
        if (!isRefreshing) {
          isRefreshing = true;
          try {
            const newSession = await refreshFn();
            if (newSession) {
              const { accessToken } = newSession;
              instance.defaults.headers.common['Authorization'] = `Bearer ${accessToken}`;
              originalRequest.headers['Authorization'] = `Bearer ${accessToken}`;
              isRefreshing = false;
            }
          } catch (err) {
            throw err;
          } finally {
            isRefreshing = false;
          }
        }

        if (!originalRequest._retry) {
          originalRequest._retry = true;
          return instance(originalRequest);
        }
      }

      if (error.response.data?.error) {
        throw CustomError.fromJSON(error.response.data.error);
      }
      throw error;
    },
  );
}
