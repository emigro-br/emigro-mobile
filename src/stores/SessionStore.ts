import * as Crypto from 'expo-crypto';
import * as SecureStore from 'expo-secure-store';
import { action, flow, makeAutoObservable, observable } from 'mobx';

import { IUserProfile } from '@/types/IUserProfile';
import { InvalidSessionError } from '@/types/errors';

import { refresh as refreshSession } from '@services/auth';
import { getUserProfile, getUserPublicKey } from '@services/emigro';

import { IAuthSession } from '../types/IAuthSession';

export class SessionStore {
  // Observable states
  justLoggedIn = false;
  session: IAuthSession | null = null;
  profile: IUserProfile | null = null;

  private authKeys: string[] = [
    'auth.accessToken',
    'auth.refreshToken',
    'auth.idToken',
    'auth.tokenExpirationDate',
    'auth.email',
    'auth.publicKey',
  ];
  private profileKey = 'user.profile';

  constructor() {
    makeAutoObservable(this, {
      // session
      session: observable,
      setSession: action,
      setPublicKey: action,
      fetchPublicKey: flow,
      // profile
      profile: observable,
      setProfile: action,
      fetchProfile: flow,
      // loggedIn
      justLoggedIn: observable,
      setJustLoggedIn: action,
    });
  }

  get accessToken() {
    return this.session?.accessToken;
  }

  get publicKey() {
    if (this.session && !this.session.publicKey) {
      this.fetchPublicKey();
    }
    return this.session?.publicKey;
  }

  setSession(session: IAuthSession | null) {
    this.session = session;
  }

  setPublicKey(publicKey: string) {
    if (this.session) {
      this.session.publicKey = publicKey;
    }
  }

  setProfile(profile: IUserProfile | null) {
    this.profile = profile;
  }

  setJustLoggedIn(justLoggedIn: boolean) {
    this.justLoggedIn = justLoggedIn;
  }

  async *fetchPublicKey() {
    console.debug('Fetching user public key...');
    const publicKey = await getUserPublicKey();
    if (this.session && publicKey) {
      this.setPublicKey(publicKey); // action will be called
      await this.save(this.session);
    }
    return publicKey;
  }

  async *fetchProfile() {
    if (this.session) {
      console.debug('Fetching user profile...');
      const profile = await getUserProfile(this.session);
      if (profile) {
        this.setProfile(profile); // action will be called
        await this.saveProfile(profile);
      }
      return profile;
    }
  }

  get isTokenExpired(): boolean {
    return !this.session ? true : this.session?.tokenExpirationDate < new Date();
  }

  async save(session: IAuthSession) {
    this.setSession(session); // FIXME: we can not replace when is only a partial update
    await Promise.all(
      Object.entries(session)
        .filter(([, value]) => value !== undefined)
        .map(([key, value]) => SecureStore.setItemAsync(`auth.${key}`, JSON.stringify(value))),
    );
  }

  async saveProfile(profile: IUserProfile) {
    // this.setProfile(profile);
    await SecureStore.setItemAsync(this.profileKey, JSON.stringify(profile));
  }

  load = async (): Promise<IAuthSession | null> => {
    const session: Partial<IAuthSession> = {};
    await Promise.all(
      this.authKeys.map(async (key) => {
        const value = await SecureStore.getItemAsync(key);
        if (value) {
          const [, attr] = key.split('.');
          session[attr as keyof IAuthSession] = JSON.parse(value);
          if (attr === 'tokenExpirationDate') {
            session.tokenExpirationDate = new Date(session.tokenExpirationDate!);
          }
        }
      }),
    );

    if (!session.accessToken) {
      return null;
    }
    if (!session.refreshToken || !session.idToken || !session.tokenExpirationDate) {
      throw new InvalidSessionError();
    }

    this.setSession(session as IAuthSession);
    this.loadProfile(); // load profile in background
    return this.session;
  };

  async loadProfile(): Promise<IUserProfile | null> {
    const profile = await SecureStore.getItemAsync(this.profileKey);
    if (profile) {
      this.setProfile(JSON.parse(profile));
    }
    return this.profile;
  }

  async clear() {
    await Promise.all(this.authKeys.map((key) => SecureStore.deleteItemAsync(key)));
    this.setSession(null);

    await SecureStore.deleteItemAsync(this.profileKey);
    this.setProfile(null);

    await this.clearPin();
    this.setJustLoggedIn(false);
  }

  signIn = async (session: IAuthSession) => {
    this.setSession(session);
    this.save(session);
    this.setJustLoggedIn(true);

    // Fetch the user data in background
    this.fetchPublicKey();
    this.fetchProfile();
  };

  async signOut() {
    await this.clear();
  }

  refresh = async () => {
    if (!this.session) {
      await this.load(); // workaround for InvalidSession when refreshing
      if (!this.session) {
        throw new InvalidSessionError();
      }
    }

    const newSession = await refreshSession(this.session);
    if (newSession) {
      newSession.publicKey = this.session.publicKey; // FIXME: workaround to avoid losing the public key on save
      await this.save(newSession);
      return this.session;
    }
  };

  async savePin(pin: string) {
    // Hash the PIN before saving it
    const hashedPin = await Crypto.digestStringAsync(Crypto.CryptoDigestAlgorithm.SHA256, pin);
    await SecureStore.setItemAsync('pin', hashedPin);
  }

  async loadPin(): Promise<string | null> {
    return await SecureStore.getItemAsync('pin');
  }

  async clearPin() {
    await SecureStore.deleteItemAsync('pin');
  }

  verifyPin = async (pin: string): Promise<boolean> => {
    const hashedPin = await this.loadPin();
    if (!hashedPin) {
      throw new Error('PIN not set');
    }

    // Hash the input PIN and compare it with the stored hashed PIN
    const inputHashedPin = await Crypto.digestStringAsync(Crypto.CryptoDigestAlgorithm.SHA256, pin);
    return hashedPin === inputHashedPin;
  };
}

export const sessionStore = new SessionStore();
