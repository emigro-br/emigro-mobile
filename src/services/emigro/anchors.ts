import { CryptoAsset } from '@/types/assets';

import { api, backendUrl } from './api';
import { InteractiveUrlRequest, InteractiveUrlResponse, Sep24Transaction } from './types';

const enum OperationType {
  DEPOSIT = 'DEPOSIT',
  WITHDRAW = 'WITHDRAW',
}

export enum CallbackType {
  CALLBACK_URL,
  EVENT_POST_MESSAGE,
}

export type ConfirmWithdrawDto = {
  transactionId: string;
  assetCode: string;
};

const getInteractiveUrl = async (
  operation: OperationType,
  params: InteractiveUrlRequest,
  callback: CallbackType,
): Promise<InteractiveUrlResponse> => {
  const endpoint = operation === OperationType.WITHDRAW ? 'withdraw' : 'deposit';
  const timeout = 15 * 1000;
  const res = await api({ timeout }).post(`/anchor/${endpoint}`, params);

  const json = res.data;
  if (callback === CallbackType.CALLBACK_URL) {
    const assetCode = params.asset_code; // FIXME: I think this is not necessary
    json.url = `${json.url}&callback=${encodeURIComponent(`${backendUrl}/anchor/withdraw-callback?assetCode=${assetCode}`)}`;
  } else if (callback === CallbackType.EVENT_POST_MESSAGE) {
    json.url = `${json.url}&callback=postMessage`;
  }

  return json;
};

export const depositUrl = async (params: InteractiveUrlRequest, callback: CallbackType) => {
  return getInteractiveUrl(OperationType.DEPOSIT, params, callback);
};

export const withdrawUrl = async (params: InteractiveUrlRequest, callback: CallbackType) => {
  return getInteractiveUrl(OperationType.WITHDRAW, params, callback);
};

export const getTransaction = async (id: string, assetCode: CryptoAsset): Promise<Sep24Transaction> => {
  const timeout = 20 * 1000;
  const res = await api({ timeout }).get('/anchor/transaction', {
    params: {
      id,
      assetCode,
    },
  });
  return res.data;
};

export const confirmWithdraw = async (data: ConfirmWithdrawDto) => {
  const timeout = 30 * 1000;
  const res = await api({ timeout }).post('/anchor/withdraw-confirm', data);
  return res.data;
};
