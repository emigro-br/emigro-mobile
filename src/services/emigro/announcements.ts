// src/services/emigro/announcements.ts
import { api } from './api';
import { AnnouncementDTO } from './types';

function normalize(a: any): AnnouncementDTO {
  return {
    id: a.id,
    title: a.title,
    description: a.description ?? null,
    imageUrl: a.imageUrl ?? a.image_url ?? null,
    deepLinkUrl: a.deepLinkUrl ?? a.deep_link_url ?? null,
    startsAt: a.startsAt ?? a.starts_at ?? null,
    endsAt: a.endsAt ?? a.ends_at ?? null,
    priority: a.priority ?? null,
    dismissed: a.dismissed ?? null,
  };
}

// Tiny helper to wait until we actually have a token (preload races)
async function waitForToken(maxMs = 3000): Promise<string | null> {
  const { sessionStore } = require('@/stores/SessionStore');
  const start = Date.now();
  while (Date.now() - start < maxMs) {
    const t = sessionStore?.accessToken;
    if (t) return t;
    await new Promise((r) => setTimeout(r, 100));
  }
  return null;
}

export async function fetchAnnouncements(): Promise<AnnouncementDTO[]> {
  const { sessionStore } = require('@/stores/SessionStore');
  const client = api();

  // Logs that tell us EXACTLY what's happening
  console.log('[announcements][FE] baseURL =', client.defaults.baseURL);
  console.log(
    '[announcements][FE] sessionStore.loaded =',
    !!sessionStore?.isLoaded,
    '| tokenLen =',
    sessionStore?.accessToken ? sessionStore.accessToken.length : 0
  );

  // Ensure we really have a token before calling (handles preload races)
  let token: string | null = sessionStore?.accessToken ?? null;
  if (!token) {
    console.log('[announcements][FE] no token yet — waiting up to 3s…');
    token = await waitForToken(3000);
  }
  console.log('[announcements][FE] using tokenLen =', token ? token.length : 0);

  try {
    console.log('[announcements][FE] GET /announcements (with explicit Authorization)');
    const res = await client.get('/announcements', {
      headers: token ? { Authorization: `Bearer ${token}` } : undefined,
    });

    console.log('[announcements][FE] status', res.status, '| type =', typeof res.data);
    if (!Array.isArray(res.data)) {
      console.log('[announcements][FE] response was not array ->', res.data);
      return [];
    }

    const out = res.data.map(normalize);
    console.log('[announcements][FE] items =', out.length);
    return out;
  } catch (err: any) {
    const s = err?.response?.status;
    const body =
      typeof err?.response?.data === 'string'
        ? err.response.data
        : JSON.stringify(err?.response?.data);
    console.warn('[announcements][FE] ERROR status =', s, '| body =', body);
    throw err; // bubble up so the store can log too
  }
}

export async function dismissAnnouncement(announcementId: string): Promise<void> {
  const { sessionStore } = require('@/stores/SessionStore');
  const client = api();
  const token = sessionStore?.accessToken ?? null;

  console.log('[announcements][FE] POST /announcements/:id/dismiss', announcementId, '| tokenLen =', token ? token.length : 0);

  try {
    const res = await client.post(
      `/announcements/${announcementId}/dismiss`,
      null,
      { headers: token ? { Authorization: `Bearer ${token}` } : undefined }
    );
    console.log('[announcements][FE] dismiss status', res.status);
  } catch (err: any) {
    const s = err?.response?.status;
    const body =
      typeof err?.response?.data === 'string'
        ? err.response.data
        : JSON.stringify(err?.response?.data);
    console.warn('[announcements][FE] dismiss ERROR status =', s, '| body =', body);
    throw err;
  }
}
