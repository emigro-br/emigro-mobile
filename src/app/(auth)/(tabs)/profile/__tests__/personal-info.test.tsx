import { render } from 'test-utils';

import { UserProfile } from '@/services/emigro/types';
import { sessionStore } from '@/stores/SessionStore';

import PersonalInfo from '../personal-info';

jest.mock('@/stores/SessionStore', () => ({
  sessionStore: {
    clear: jest.fn(),
    publicKey: 'test-public',
    get profile() {
      return null;
    },
  },
}));

const profileMock = {
  given_name: 'Test Name',
  family_name: 'Test Last Name',
  email: 'test@email.com',
  address: 'Test Address',
} as UserProfile;

describe('PersonalInfo screen', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('Should appear loading screen while fetching the user information', () => {
    jest.spyOn(sessionStore, 'profile', 'get').mockReturnValue(null);
    const { getByTestId } = render(<PersonalInfo />);

    expect(getByTestId('loading-spinner')).toBeOnTheScreen();
  });

  test('Should render the Profile screen correctly', () => {
    jest.spyOn(sessionStore, 'profile', 'get').mockReturnValue(profileMock);
    const screen = render(<PersonalInfo />);
    expect(screen.toJSON()).toMatchSnapshot();
  });
});
