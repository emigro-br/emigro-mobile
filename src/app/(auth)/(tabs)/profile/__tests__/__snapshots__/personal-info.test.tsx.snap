// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonalInfo screen Should render the Profile screen correctly 1`] = `
<View
  style={
    [
      {},
      {
        "flex": 1,
        "height": "100%",
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <View
    className="flex-1 bg-white"
  >
    <View
      className="flex-col gap-4 p-4"
    >
      <View>
        <Text
          className="font-body text-base text-typography-500"
        >
          Full Name
        </Text>
        <Text
          className="text-typography-700 font-body text-lg"
        >
          Test Name Test Last Name
        </Text>
      </View>
      <View />
      <View>
        <Text
          className="font-body text-base text-typography-500"
        >
          Email
        </Text>
        <Text
          className="text-typography-700 font-body text-lg"
        >
          test@email.com
        </Text>
      </View>
      <View />
      <View>
        <Text
          className="font-body text-base text-typography-500"
        >
          Address
        </Text>
        <Text
          className="text-typography-700 font-body text-base"
        >
          Test Address
        </Text>
      </View>
      <View />
    </View>
  </View>
</View>
`;
